artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/connect/mirror/AMQPMirrorControllerSource.java

```diff
@@ -0,0 +1,241 @@
+import java.util.Map;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.PooledByteBufAllocator;
+import org.apache.activemq.artemis.api.core.Message;
+import org.apache.activemq.artemis.api.core.QueueConfiguration;
+import org.apache.activemq.artemis.api.core.SimpleString;
+import org.apache.activemq.artemis.core.paging.PagingStore;
+import org.apache.activemq.artemis.core.server.ActiveMQComponent;
+import org.apache.activemq.artemis.core.server.ActiveMQServer;
+import org.apache.activemq.artemis.core.server.MessageReference;
+import org.apache.activemq.artemis.core.server.Queue;
+import org.apache.activemq.artemis.core.server.RoutingContext;
+import org.apache.activemq.artemis.core.server.impl.AckReason;
+import org.apache.activemq.artemis.core.server.impl.AddressInfo;
+import org.apache.activemq.artemis.core.server.impl.RoutingContextImpl;
+import org.apache.activemq.artemis.core.server.remotecontrol.MirrorController;
+import org.apache.activemq.artemis.core.transaction.Transaction;
+import org.apache.activemq.artemis.protocol.amqp.broker.AMQPMessage;
+import org.apache.activemq.artemis.protocol.amqp.broker.AMQPStandardMessage;
+import org.apache.activemq.artemis.protocol.amqp.util.NettyWritable;
+import org.apache.activemq.artemis.protocol.amqp.util.TLSEncode;
+import org.apache.qpid.proton.amqp.Symbol;
+import org.apache.qpid.proton.amqp.messaging.AmqpValue;
+import org.apache.qpid.proton.amqp.messaging.DeliveryAnnotations;
+import org.apache.qpid.proton.amqp.messaging.Header;
+import org.apache.qpid.proton.amqp.messaging.MessageAnnotations;
+import org.apache.qpid.proton.amqp.messaging.Properties;
+import org.apache.qpid.proton.amqp.messaging.Section;
+import org.apache.qpid.proton.codec.EncoderImpl;
+import org.apache.qpid.proton.codec.WritableBuffer;
+import org.jboss.logging.Logger;
+
+public class AMQPMirrorControllerSource implements MirrorController, ActiveMQComponent {
+
+   private static final Logger logger = Logger.getLogger(AMQPMirrorControllerSource.class);
+
+   public static final Symbol EVENT_TYPE = Symbol.getSymbol("ma.EVENT_TYPE");
+   public static final Symbol ADDRESS = Symbol.getSymbol("ma.ADDRESS");
+   public static final Symbol QUEUE = Symbol.getSymbol("ma.QUEUE");
+
+   // Events:
+   public static final Symbol ADD_ADDRESS = Symbol.getSymbol("addAddress");
+   public static final Symbol DELETE_ADDRESS = Symbol.getSymbol("deleteAddress");
+   public static final Symbol CREATE_QUEUE = Symbol.getSymbol("createQueue");
+   public static final Symbol DELETE_QUEUE = Symbol.getSymbol("deleteQueue");
+   public static final Symbol ADDRESS_SCAN_START = Symbol.getSymbol("AddressCanStart");
+   public static final Symbol ADDRESS_SCAN_END = Symbol.getSymbol("AddressScanEnd");
+   public static final Symbol POST_ACK = Symbol.getSymbol("postAck");
+
+   // Delivery annotation property used on remote control routing and Ack
+   public static final Symbol INTERNAL_ID = Symbol.getSymbol("ma.INTERNAL_ID");
```

Comment:
The problem with using Message Annotation is that I would require to re-encode the message, since the messageID is generated after the storage on the broker.

I need a property that's only changed during delivery, and delivery annotation seemed to be the best usage I could find.

We would need to have intermediaries copying this delivery annotation to have this eventually working.

Reply:
Intermediaries are not going to copy delivery annotations, they are specifically not meant to. Thats the main distinction between the two types, i.e why both exist.

I dont see that handling the message annotations would be much differnet in terms of skipping sending bytes than what youve already added for the delivery annotations. Key difference, also add the recieved message annotation to the ones that are there (which isitself kind of similar to what my other comments are implying around any other explicitly-set delivery annotations using the previous DA mechanism)